node/network/bitfield-distribution/src/lib.rs

```diff
@@ -0,0 +1,684 @@
+where
+	Context: SubsystemContext<Message = BitfieldDistributionMessage>,
+{
+
+	// notify the overseer about a new and valid signed bitfield
+	ctx.send_message(
+		AllMessages::Provisioner(
+			ProvisionerMessage::ProvisionableData(
+				ProvisionableData::Bitfield(
+					message.relay_parent.clone(),
+					message.signed_availability.clone(),
+				)
+			)
+		)
+	).await;
+
+	let message_sent_to_peer = &mut (job_data.message_sent_to_peer);
+
+	// concurrently pass on the bitfield distribution to all interested peers
+	let interested_peers = peer_views
+		.iter()
+		.filter_map(|(peer, view)| {
+			// check interest in the peer in this message's relay parent
+			if view.contains(&message.relay_parent) {
+				// track the message as sent for this peer
+				message_sent_to_peer
+					.entry(peer.clone())
+					.or_default()
+					.insert(validator.clone());
+
+				Some(peer.clone())
+			} else {
+				None
+			}
+		})
+		.collect::<Vec<PeerId>>();
+
+	ctx.send_message(AllMessages::NetworkBridge(
+		NetworkBridgeMessage::SendMessage(
+			interested_peers,
+			BitfieldDistribution::PROTOCOL_ID,
+			Encode::encode(&message),
+		),
+	))
+	.await?;
+	Ok(())
+}
+
+/// Handle an incoming message from a peer.
+async fn process_incoming_peer_message<Context>(
+	ctx: &mut Context,
+	tracker: &mut Tracker,
+	origin: PeerId,
+	message: BitfieldGossipMessage,
+) -> SubsystemResult<()>
+where
+	Context: SubsystemContext<Message = BitfieldDistributionMessage>,
+{
+	// we don't care about this, not part of our view
+	if !tracker.view.contains(&message.relay_parent) {
+		return modify_reputation(ctx, origin, COST_NOT_INTERESTED).await;
+	}
+
+	// Ignore anything the overseer did not tell this subsystem to work on
+	let mut job_data = tracker.per_relay_parent.get_mut(&message.relay_parent);
+	let job_data: &mut _ = if let Some(ref mut job_data) = job_data {
+		job_data
+	} else {
+		return modify_reputation(ctx, origin, COST_NOT_INTERESTED).await;
+	};
+
+	let validator_set = &job_data.validator_set;
+	if validator_set.len() == 0 {
+		return modify_reputation(ctx, origin, COST_MISSING_PEER_SESSION_KEY).await;
+	}
+
+	// use the (untrusted) validator index provided by the signed payload
+	// and see if that one actually signed the availability bitset
+	let signing_context = job_data.signing_context.clone();
+	let validator_index = message.signed_availability.validator_index() as usize;
+	let validator = if let Some(validator) = validator_set.get(validator_index) {
+		validator.clone()
+	} else {
+		return modify_reputation(ctx, origin, COST_VALIDATOR_INDEX_INVALID).await;
+	};
+
+	if message
+		.signed_availability
+		.check_signature(&signing_context, &validator)
+		.is_ok()
+	{
+		let one_per_validator = &mut (job_data.one_per_validator);
+		// only relay_message a message of a validator once
+		if one_per_validator.get(&validator).is_some() {
+			trace!(target: "bitd", "Alrady received a message for validator at index {}", validator_index);
+			return Ok(());
```

Comment:
Here, we have to be careful. Is it true that, if we're already aware of a bitfield for a validator, this branch will always be hit?

If so, then a peer could flood us with bitfields that we'd do signature-checking work on, but drop immediately.

However, we shouldn't unconditionally reduce the reputation of the peer - if they want to send us a bitfield and just aren't the first one to get there, we shouldn't punish them for that.

One way to deal with this is to track which validators a peer has _sent us_ bitfields for (per relay-parent). If this is the first one, we reward them and return `Ok`. If not, then they are flooding us and we punish them and return `Ok`.

Reply:
Reusing `message_for_validator_needed_by_peer` is racy, because you might have a bitfield from the peer->you in-flight at the moment we send them a bitfield. This would lead us to punish them on receipt of the first bitfield sent to us.